core/src/main/java/hudson/PluginWrapper.java

```diff
@@ -394,6 +426,9 @@ public PluginWrapper(PluginManager parent, File archive, Manifest manifest, URL
 		this.active = !disableFile.exists();
 		this.dependencies = dependencies;
 		this.optionalDependencies = optionalDependencies;
+        for (Dependency d : optionalDependencies) {
+            assert d.optional : d + " included among optionalDependencies of " + shortName + " but was not marked optional";
```

Comment:
Unfortunately there was duplicated information here (a field and also membership in a collection) which had drifted out of synch.